### ISSUE TITLE
Android Pay Support?

### ISSUE THREAD

**Issue by ShapeShifter499** [microg/android_packages_apps_GmsCore#361]
I'm trying to use Android Pay but it constantly crashes with the following error message. "Android Pay is unavailable. Android Pay isn't available"   I know I pass SafetyNet because I'm able to play Pokemon Go without issue. Is it possible to add in support for Android Pay?

I've taken a logcat during the crash. 
[android_pay_logcat.txt](https://github.com/microg/android_packages_apps_GmsCore/files/923482/android_pay_logcat.txt)

System
Moto X Pure
LineageOS 14.1
Magisk 12
microG 0.2.4-81-gb1e98ea

**Comment by mar-v-in** [comment]
The service bound by `com.google.android.gms.tapandpay.service.BIND` is not yet implemented and likely required for Android Pay to work.

**Comment by ShapeShifter499** [comment]
@mar-v-in Would this be hard to implement? Or even legal since this is a payment service?

**Comment by ShapeShifter499** [comment]
@mar-v-in I use microG for a mix of security, privacy, and battery saving so I'd like to see support for Android Pay implemented.   I've added $5 to a bounty

**Comment by ShapeShifter499** [comment]
Any potential updates on this?

**Comment by ShapeShifter499** [comment]
Added another 5$ to this in hope someone can add support for this.

**Comment by ShapeShifter499** [comment]
@mar-v-in any chance of Android Pay support ever being implemented?

**Comment by ShapeShifter499** [comment]
Is this being ignored for some reason? Will this be hard to implement?

**Comment by youcanaddthenameofyourorganizationlater** [comment]
I too would like to see support for this.

**Comment by ale5000-git** [comment]
@ShapeShifter499: It is only one developer so if it work on one thing all others are ignored, it will be faster if there will be more contributors.

**Comment by ShapeShifter499** [comment]
@ale5000-git I wish I knew more to help with the code but I barely know code at all, just simple scripts.

**Comment by ale5000-git** [comment]
To add money for this issue the link is [this](https://www.bountysource.com/issues/44168158-android-pay-support).

**Comment by ShapeShifter499** [comment]
I bumped up the bounty to a total of $25

**Comment by jreinert** [comment]
Bumped to $50

**Comment by jacen05** [comment]
Bumped to 100$. 
This feature would be very useful indeed.

**Comment by Banginpadr** [comment]
i could try to help but to be honest i dont have an idea of what ya trying to get done here.

**Comment by ShapeShifter499** [comment]
@Banginpadr   @mar-v-in suggested that it might work if a few things were implemented in microG.  It would be nice to have some control over what Google does with the device by installing microG but then allow some aspects to work like tap-to-pay with Google Pay. 

It's likely your transactions are heavily tracked as it is. Some people, such as I, may not care if Google knows they brought groceries somewhere at sometime.

**Comment by Banginpadr** [comment]
> @Banginpadr @mar-v-in suggested that Google Pay might work if a few things were implemented in microG. It would be nice to have some control over what Google does with the device by installing microG but then allow some aspects to work like tap-to-pay with Google Pay.
> 
> It's likely your transactions are heavily tracked as it is. Some people, such as I, may not care if Google knows they brought groceries somewhere at sometime. Cash is the only way to go if you're really paranoid.

i guess you need to use api for that to work.

**Comment by Axolord** [comment]
This issue can't be fixed, because SafetyNet does not pass in more modern devices, if the bootloader is unlocked. Furthermore, SafetyNet itself is not passable in microG at the moment, like at all. Even the older devices (that do pass SafetyNet with unlocked booloader under Gapps) do not pass.

It is only a matter of time until SafetyNet won't be passable on every device with an unlocked bootloader.
Source: https://www.xda-developers.com/magisk-no-longer-hide-bootloader-unlock-status/

So in my opinion it would be a waste of time to implement Android Pay, just to have it not working again in a year or less time - again.

correct me if I'm wrong,
cheers

**Comment by notpushkin** [comment]
> SafetyNet does not pass in more modern devices, if the bootloader is unlocked

Can't we simulate an older device instead? I know all the dance around SafetyNet is tricky and will continue to become even trickier over years, but still.

**Comment by TheOneWithTheBraid** [comment]
Well, isn't it possible to hide both unlocked bootloader and root using Magisk and the Riru EdXposed HiddenCore module? I tried so on Android 11. At least, using the proprietary GApps, Google Pay works this way. This should hence similarly work with proper microG implementations too.

**Comment by Kranzes** [comment]
I am not sure that I am understanding this correctly.
Let's say I have a device like one of the newer Redmi's from Xiaomi, with MicroG, Magisk (magisk hide), but without any GAPPS,
Will I be able to use Google Pay with it? Now with the new flutter based Google Pay app.

**Comment by TheLastGimbus** [comment]
Hi there! My banking app (mBank, `pl.mbank`) recently added their own nfc payemnts (*without* google pay) - it looks like everything is going well, and the app even opens when I touch terminal with my phone (!!!) but then it fails

(I can provide more details later of course)

Would `com.google.android.gms.tapandpay.service.BIND` be hard to implement? Is it just some "launch X when Y intent received" or some more API logic etc? If I got some guidance, I could try...

**Comment by Kranzes** [comment]
I have to add that I am using tap-and-pay (contact-less payment via NFC) using my bank's app (Leumi Bank).
I also only have MicroG without any GAPPS, everything in the app works fine.

**Comment by notpushkin** [comment]
Some Belarus banks have dedicated apps that also work pretty well with tap and pay.

In Russia, we also have an app called [Wallet](https://cardsmobile.ru/en/) (called Koshelek in Russian; not to be confused with Apple Wallet) which many different banks (in both 🇷🇺 and 🇧🇾, and they plan on expanding to 🇪🇺 next year I think) support in addition to Google Pay. Sadly, it doesn't seem work on my microG setup, but it might be interesting to explore.

**Comment by Kranzes** [comment]
Making Google Pay work with MicroG would be fantastic, though, I don't know if it will work because of SafetyNet.

**Comment by rubikscraft** [comment]
@Kranzes that is not a problem, I've already checked and the new flutter based google pay does not check for safetynet. The older one does tho, but that one is discontinued so it doesnt matter.

**Comment by PylotLight** [comment]
> 

I've tried the latest google play and it just spins on launch with a small white rectangle, are you saying you've gotten further than that?

**Comment by rubikscraft** [comment]
If you have a small white rectangle you have the old pay version (https://play.google.com/store/apps/details?id=com.google.android.apps.walletnfcrel). 
With the new version you should get dropped into a full screen app immediately: (https://play.google.com/store/apps/details?id=com.google.android.apps.nbu.paisa.user)

**Comment by rubikscraft** [comment]
If you are using aurora store you should be able to open those links. You should also be able to change your google play country with it. Maybe you can get the new one that way. Or you could just download the apk from apkmirror: <https://www.apkmirror.com/apk/google-inc/tez-a-new-payments-app-by-google/>.
Either way, region locks are stupid.

**Comment by rubikscraft** [comment]
Oh and in case you didnt realize, the new app only does not check for safetynet anymore. It still requires all the other pay intents that have not yet been implemented into microg. 
So if you trying to get it working with microg, right now it wont work. But if you are trying to get it working on a device with gapps but a broken safetynet, the new app should fix it.

**Comment by dumprop** [comment]
> Some Belarus banks have dedicated apps that also work pretty well with tap and pay.
> 
> In Russia, we also have an app called [Wallet](https://cardsmobile.ru/en/) (called Koshelek in Russian; not to be confused with Apple Wallet) which many different banks (in both ru and belarus, and they plan on expanding to eu next year I think) support in addition to Google Pay. Sadly, it doesn't seem work on my microG setup, but it might be interesting to explore.

MIR Pay should work

**Comment by dylangerdaly** [comment]
I'd like to see this implemented, I'd prefer however if I could hand over my Google creds to GooglePay specifically, and not microG/all other Google apps I have installed.

I don't know if that's possible.

**Comment by puuuuh** [comment]
> > Some Belarus banks have dedicated apps that also work pretty well with tap and pay.
> > In Russia, we also have an app called [Wallet](https://cardsmobile.ru/en/) (called Koshelek in Russian; not to be confused with Apple Wallet) which many different banks (in both ru and belarus, and they plan on expanding to eu next year I think) support in addition to Google Pay. Sadly, it doesn't seem work on my microG setup, but it might be interesting to explore.
> 
> MIR Pay should work

It doesnt, because google sign-in not implemented yet:)

**Comment by dumprop** [comment]
> > > Some Belarus banks have dedicated apps that also work pretty well with tap and pay.
> > > In Russia, we also have an app called [Wallet](https://cardsmobile.ru/en/) (called Koshelek in Russian; not to be confused with Apple Wallet) which many different banks (in both ru and belarus, and they plan on expanding to eu next year I think) support in addition to Google Pay. Sadly, it doesn't seem work on my microG setup, but it might be interesting to explore.
> > 
> > 
> > MIR Pay should work
> 
> It doesnt, because google sign-in not implemented yet:)

Hm, maybe.. I'm not sure because maybe I had the original Google services in the sandbox

**Comment by dumprop** [comment]
> > > Some Belarus banks have dedicated apps that also work pretty well with tap and pay.
> > > In Russia, we also have an app called [Wallet](https://cardsmobile.ru/en/) (called Koshelek in Russian; not to be confused with Apple Wallet) which many different banks (in both ru and belarus, and they plan on expanding to eu next year I think) support in addition to Google Pay. Sadly, it doesn't seem work on my microG setup, but it might be interesting to explore.
> > 
> > MIR Pay should work
> 
> It doesnt, because google sign-in not implemented yet:)

A year ago, the mir pay did not require Google services. Now such a problem, yes

**Comment by dada513** [comment]
@TheLastGimbus does mBank work now? I am considering installing microG but I need my bank app

**Comment by TheLastGimbus** [comment]
mBank, as is, works very well 💯 

But Blik nfc doesn't :/ but blick through code - perfectly fine 👍

**Comment by droserasprout** [comment]
Thanks for your participation, guys! I suggest creating a GitHub Wiki page in this project to gather reports about the current status of various payment services with microG installed.

**Comment by dylangerdaly** [comment]
I think Google dropped the SafetyNet requirement for GooglePay, so it should be a little easier to suport it

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by marckrieger** [comment]
I just passed SafetyNet with LineageOS 18.1 for MicroG and the GooglePay app still didn't work (inmediately crashes and a loading pop-up shows up).

**Comment by rubikscraft** [comment]
@marckrieger are you sure you are using the new flutter based google pay app?
The loading pop-up you are talking about sounds more like the older version. The app id should be `com.google.android.apps.nbu.paisa.user` for the new version.

**Comment by marckrieger** [comment]
@rubikscraft you are right it wasn't the flutter version. I tried with this one and it doesn't work either. It just gets stuck with the splash screen.

**Comment by rubikscraft** [comment]
@marckrieger Hmm, weird, but since you passed saftenet I suppose it checks for some other things. You are running normal Google play services right? Since right now Google Pay does not work with microg.

**Comment by marckrieger** [comment]
No, I am using MicroG, I know it's normal it doesn't work. I just wanted to share that GPay does not work with MicroG even with passing SafetyNet, in case it was not clear yet.

**Comment by MagicRB** [comment]
> @marckrieger Hmm, weird, but since you passed saftenet I suppose it checks for some other things. You are running normal Google play services right? Since right now Google Pay does not work with microg.

Micro is missing APIs which google pay relies on and assumes are present. No need to check, since it's not supposed to be installed on a non-Googled device.

**Comment by MagicRB** [comment]
> Micro is missing APIs which google pay relies on and assumes are present. No need to check, since it's not supposed to be installed on a non-Googled device.

Id really love to know why my comment was marked outdated, if what I said is incorrect I apologize and feel free to correct. But just marking as outdated is lame.

**Comment by steewbsd** [comment]
I'd also like to see if someone could take a look at this. Shame it has been halted.

**Comment by tslpre** [comment]
And by the way, if one cannot use the flutter version of the Google Pay (com.google.android.apps.nbu.paisa.user), like me, as it does not covers Spain, then you have to install the old one, and it gets stuck at a little white rectangle when loading it.

**Comment by notpushkin** [comment]
Let's raise the bar a bit. I'm not really comfortable with using Bountysource for this, but if anybody wanna work on bringing this in after all these years, **I can chip in another $285** in addition to that. Only if it works on my device though :^)

**Comment by ustas-eth** [comment]
Let's dig up this old thread :)
Now that SafetyNet is easily passable (and it seems not required to run Android Pay), I tried to install the app, and it sent me to the Microg's Google login page. After, I tried to open the app once again, but the API of Microg seemed to miss something:

![Screenshot_20230703-085810_microG Services Core](https://github.com/microg/GmsCore/assets/82833595/ba39abfe-2cdc-4368-9189-e1d7bb1552f9)

**Comment by ustas-eth** [comment]
> @ustas-eth
> 
> > Now that SafetyNet is easily passable
> 
> I might have missed something but this sounds like a pretty bold claim to me. Care to share any details on what changed?

Sure!
Actually, nothing changed. The only things that are needed are root, Magisk, and the latest version of modded Universal SafetyNet Fix by

**Comment by notpushkin** [comment]
I see, thank you!

<details>
<summary>Slightly off-topic...</summary>

> Ideally, this workaround should be incorporated in custom ROMs instead of injecting code with a Magisk module. \
> _– https://github.com/Displax/safetynet-fix/tree/dev#rom-integration_

I wonder if this might be a good patch to include in [LineageOS for microG](https://lineage.microg.org/) 🤔 

</details>

**Comment by rokejulianlockhart** [comment]
@ustas-eth, I believe that SafetyNet remains necessary to use Pay, because I run my phone vendor's stock ROM with merely an unlocked bootloader, and yet can't use Pay.

**Comment by KCGD** [comment]
I've managed to get somewhat far into google pay with my current setup. I was able to select an account and enter my phone number, but I always get met with the "this device cannot run google pay" error message. I'm current passing all safetynet checks consistently.
Here's a full report of all of the frameworks, modules and testing apps i have installed, as well as their configurations:

- Magisk 26.3 (using magisk's app spoofing) with zygisk installed
- Magisk modules:
- - Hide UserDebug, Test-Keys and LineageOS [v1]
- - MagiskHide Props Config [v6.1.2-v137]
- - microG Installer revived [3.3.2-0]
- - SafetyNet sensitive props [v3]
- - Shamiko [v0.7.3]
- - Systemless Hosts (built-in) [v1.0]
- - Universal SafetyNet Fix (Displax mod) [v2.4.0-MOD_1.3-migroG]
- - Zygisk - LSPosed [v1.9.1 (6990)]
- LSPosed (Via Magisk)
- LSPosed modules:
- - FakeGapps [v4.2]
- - Hide My Applist [v3.2]

I've set up some spoofing / app hiding apps to help pass other security checks
- Shizuku
- Amarok (using Shizuku-based app hiding)
- Apps on my amarok hide list:
- - Magisk (the spoofed app)
- - LSPosed
- - Hide My Applist
- Hide My Applist
- Apps hidden on Hide My Applist
- - Every app that has root access or can access root in ANY capacity (including root checkers)
- - Every app that can interact with Xposed or LSPosed
- - Every microG related app
- These apps are hidden from any app that could possibly look for them (all banking apps, flight apps, payment apps)

Here's the apps I use to test my root hiding / test the security I have set up:
- Ruru
- - All check pass except:
- - Abnormal environment:
- - - Su File found
- - - Magisk file found
- - - Xposed Edge recieved "suspicious" marker
- - PM Sundry APIs
- - - org.lsposed.manager found
- - - com.tsng.hidemyapplist found
- - Xposed Modules
- - - UniversalAuth found (not related to security, i just use for facial unlock)
- - - FakeGapps found
- - Accessibility Services and Settings Props both recieved "suspicious", but it was unrelated to anything i've previously mentioned
- Oprek Detector
- - All tests passed EXCEPT Magisk Module Indications (Oprek still reported device healthy status)
- Applist Detector
- - All tests passed.
- Root and SafetyNet Checker by BetterOpts
- - All tests passed
- - - SafetyNet request Passed
- - - CTS profile match Passed
- - - Basic integrity Passed
- SafetyNet Checker by flinkapps
- - All tests passed
- SafetyNet Helper Sample by ScottyAB Apps
- - inconclusive: Recieved Response signature validation error
- Play Integrity API Checker by Nikolas Spirdakis
- - Errored: Couldnt get token from google (Binding to the service in the Playstore failed)
- MicroG's safetynet test
- - Passed all tests

All tests were run with Amarok's app hiding on, and all the listed modules enabled.
Note: google pay will show a white screen unless running off of a fresh install of the app.

**Comment by ustas-eth** [comment]
Good day, everyone! Let me remind, this issue is still very much alive

**Comment by invaliddayta** [comment]
I'd like to support this but the bounty link does not seem to be valid anymore

**Comment by LeVraiRoiDHyrule** [comment]
I would be very interested in any progression made on this, it probably is one of the biggest challenges for microg

**Comment by MrNessuno** [comment]
I tryed to install Android pay (2017 version app), gpay (2022) and Google wallet (latest version) but with all the apps can't find my Google account. So I'm stuck even before opening the app.
I checked in Setting under the Google account the "enable app to discover the Google account" option.
Screenshot:
![Screenshot_20231111-164404_Google Wallet](https://github.com/microg/GmsCore/assets/20603088/107b3041-9cbf-43eb-aae9-b0b4a56d331d)

**Comment by Iey4iej3** [comment]
There was some progress: https://github.com/microg/GmsCore/commit/91e830b102b8a46d296c9f338ab46cfc8522b95f

**Comment by KCGD** [comment]
> There was some progress: [91e830b](https://github.com/microg/GmsCore/commit/91e830b102b8a46d296c9f338ab46cfc8522b95f)

Yay!! It looks like they're laying the groundwork for pay support. It isn't close to functional yet though.

**Comment by Revyrun607** [comment]
Any way we can still add money to the bounty thing? I'd really appreciate this being available HW devices

**Comment by notpushkin** [comment]
I'm afraid Bountysource has scammed everybody: https://github.com/bountysource/core/issues/1539

My [$500 offer](https://github.com/microg/GmsCore/issues/361#issuecomment-1347431843) still stands though :-)

**Comment by ignisf** [comment]
Guys, would you genuinely risk something automatically flagging your Google account for fraud because of using payments without Google's blessing?

I mean, the rest of microg they may be OK with ignoring... However when payments are concerned, there might be risk of *outside regulators* having an issue with Google, which I expect would warrant a swift reaction.

And while microg may not be affected by direct legal action or could be protected by interoperability clauses, if Google bans all microg users that dare to use payments and word goes out, that's it with the project...

**Comment by KCGD** [comment]
Personally, I don't think this would be an issue. All payments would still go through Google and be appropriately charged to whatever account they're linked to. MicroG's implementation will allow Google pay to work as intended. 

I don't believe any sort of fraud would be possible either, and Googles servers would still need to process and approve this request before any charges are made. 

Also, even if payment does not work for some reason, Google pay is still used for other things like nfc based ID cards, which would also be awesome to have work. 

If microG runs into legal trouble down the road, we can revert pay support and just he back where we are now, but I'd say it's worth a shot.

**Comment by CutestNekoAqua** [comment]
How can we contribute to this? I'd like to help where I can, cause this is the only thing which will make me need to switch back to normal google services for the time being, cause I need google pay :p 

Might also be interesting for other people looking to help, so just pop some info here if you have it ;)

**Comment by Kaydax** [comment]
I opened up a new bounty for this issue if anyone is interested in helping fund this: https://gitpay.me/#/task/797

**Comment by rugk** [comment]
@Kaydax Yet on another platform that may inevitable scam you? (or be bought etc.) [After the bountysource issue](https://github.com/microg/GmsCore/issues/361#issuecomment-1929330530), I am not sure people will built trust in such platforms again.

**Comment by alexanmtz** [comment]
@rugk @Kaydax I'm a maintainer of Gitpay, which is entirely Open Source, and I've been supporting the project for more than six years. I can assure you we are not a scam, and we're not under any big company umbrella to dictate our direction; it is made by developers for developers, helping developers receive bounties across the globe, and we do not retain the funds who merge A PR to this issue can claim the bounty from your payments.

I would love to have you on board, and any feedback is very welcomed 🙏

**Comment by knocte** [comment]
> I opened up a new bounty for this issue if anyone is interested in helping fund this: https://gitpay.me/#/task/797

Hey @Kaydax did you add any amount to it? It looks empty.
With regards to funding options, I'm very sad about BountySource's rugpull, but happy that there are still offers here from some comments, such as @notpushkin's. As for me, I'd be willing to contribute money too, but only with crypto. To this end, I think the platform that may take BountySource/Gitcoin's baton might be SphinxBounties system: https://community.sphinx.chat/bounties But I haven't tested it yet (I don't see that GitPay supports crypto btw, that's why I'm discarding it as an idea)

With regards to this issue itself, I guess its title should be renamed? AndroidPay was the name of Google's Android's option to pay, but that changed to GooglePay as far as I understand, and these days it's Google Wallet actually.

In fact, to share my 2 sats, I bought a Murena phone recently (hardware: Fairphone5) that has e/OS 2.0 pre-installed (in particular, 2.0-t-20240506399547-stable-FP5) and this was the result when trying to run GoogleWallet:
- Had microG with no Google accounts.
- Installed Google Wallet from AppLounge.
- Run Google Wallet: it asks me to log in into my Google account.
- I log into my Google Account after passing the 2FA test(s).
- If I enter microG settings I can see now that my google account is listed in the Accounts section.
- Running GoogleWallet seems to sort of work, however when I click on the button "+ Add to Wallet", I get this error message:
"Not yet supported: com.google.android.gms.pay.homescreen.additem.ADD_ITEM"

Regards to all, especially opensource supporters&contributors.

**Comment by CutestNekoAqua** [comment]
Anything we can help to get this to work?

**Comment by knocte** [comment]
> Anything we can help to get this to work?

Hi @CutestNekoAqua: no idea, I'm a dev but not Android dev. From what I can see, there's a new linked issue to this issue: https://github.com/microg/GmsCore/issues/2086 where someone claims that their microG version is not new enough. I myself have checked in my phone (e/OS 2.1-t-xyz) and I have a version that is too old as well (0.3.1.x). Apparently 0.3.2.y may improve things. Are you able to test on your end and try to upgrade to that microG version?

**Comment by Iey4iej3** [comment]
The MicroG (0.3.2.240913) that I installed does not seem to have anything working about Google Pay as well.

**Comment by knocte** [comment]
@Iey4iej3 please report that in #2086

**Comment by nociza** [comment]
Mostly wondering if this will support for hw pay as well? Microg is already a godsend for making hw's useable right now

**Comment by knocte** [comment]
> Mostly wondering if this will support for hw pay as well? Microg is already a godsend for making hw's useable right now

Excuse me, what does 'hw' mean exactly?

**Comment by notpushkin** [comment]
> Excuse me, what does 'hw' mean exactly?

Huawei, perhaps?

**Comment by LeVraiRoiDHyrule** [comment]
Hi,
I guess that Google Pay/Wallet is the final boss of microG's journey, and that it will probably take a lot of time before someone has the ressources and time to invest in this.

While we wait, I wonder if there would be a way to have google pay on a microG phone by having a containerized version of the official Google Play Services. Would there be a way to have a container / virtual machine or similar, on the phone, that would contain only Google Wallet and official Google Play Services ? That way, maybe we could limit Play Services tracking while allowing to use Google Wallet.

Is there such thing as a containerized Play Services that could be installed along microG ?

Thanks in advance for any answer.

**Comment by Weissnix4711** [comment]
> Hi, I guess that Google Pay/Wallet is the final boss of microG's journey, and that it will probably take a lot of time before someone has the ressources and time to invest in this.
> 
> While we wait, I wonder if there would be a way to have google pay on a microG phone by having a containerized version of the official Google Play Services. Would there be a way to have a container / virtual machine or similar, on the phone, that would contain only Google Wallet and official Google Play Services ? That way, maybe we could limit Play Services tracking while allowing to use Google Wallet.
> 
> Is there such thing as a containerized Play Services that could be installed along microG ?
> 
> Thanks in advance for any answer.

GrapheneOS does this, though I've never tried it because I don't have a supported device. I don't think any other ROM does a similar thing.

**Comment by notpushkin** [comment]
DivestOS used to do this for both Google Play and microG, although nowadays I think they support microG only.

<details><summary><strong>Offtop rant</strong></summary>
I'm using this now and while it's neat and secure, it doesn't support UnifiedNlp at all for political reasons (“The primary reason for not supporting network location providers is that they effectively divulge your location to third-parties (eg. Google/Qualcomm/Apple/Mozilla) every time they are used”). Then again, they are working on [an UnifiedNlp backend](https://codeberg.org/divested-mobile/wlanbackend), too, so maybe that won't always be the case.
</summary>

**Comment by yangyang628** [comment]
Is there any new progress on how to use Google Wallet?

**Comment by znbaboy** [comment]
hopefully this is solved soon. this is the only thing holding me back from switching to MicroG permanently

**Comment by robog-two** [comment]
Even if it's not possible to get say, debit cards to work, due to the restrictions on SafetyNet. I would appreciate if it were possible to get loyalty cards (A great example would be a CVS card) or other passes that are just a number & barcode. I don't expect 100% compatibility with normal play services, but I think the basic framework/UI for the card system would be really nice.

**Comment by brennoflavio** [comment]
> Even if it's not possible to get say, debit cards to work, due to the restrictions on SafetyNet. I would appreciate if it were possible to get loyalty cards (A great example would be a CVS card) or other passes that are just a number & barcode. I don't expect 100% compatibility with normal play services, but I think the basic framework/UI for the card system would be really nice.

Have you tried Catima? Its a free app that should cover this use case

**Comment by vishwamartur** [comment]
still open?

**Comment by KCGD** [comment]
I believe so. I've seen some progress in the Google Wallet app. It no longer shows the "not implemented" error, but the UI never shows any meaningful data at this point. ![Image](https://github.com/user-attachments/assets/d927ecd1-88f6-4503-bb7b-d1f435ef814a)

**Comment by KaKi87** [comment]
> Google Pay is a proprietary service developed by Google, and as such, it's unlikely to be directly implemented in MicroG

Isn't the purpose of MicroG to make Google Apps see it as genuine Google Services ?

With Play Integrity now implemented, Google Wallet compatibility should be possible.

**Comment by ale5000-git** [comment]
> Google Pay is a proprietary service developed by Google, and as such, it's unlikely to be directly implemented in MicroG, which is an open-source alternative to Google's proprietary services. Unless it will integrate with some other kind of digital wallet.. probably open wallet foundation is a good start

Let's not get into useless discussions, almost nothing is impossible.
Let's stay in a constructive discussion.
Even SafetyNet and GCM are a sort of "proprietary services" but now they works.

**Comment by dylangerdaly** [comment]
> With Play Integrity now implemente

Wat? Can I pls have a link

-------- Mensagem original --------

**Comment by Iey4iej3** [comment]
> With Play Integrity now implemente
> Wat? Can I pls have a link

https://github.com/microg/GmsCore/releases/tag/v0.3.6.244735

**Comment by notpushkin** [comment]
This is amazing! Best Christmas gift this year :-) 

Any idea what’s left for Pay / Wallet support?

**Comment by xusism** [comment]
Hi, realll

> > Google Pay is a proprietary service developed by Google, and as such, it's unlikely to be directly implemented in MicroG, which is an open-source alternative to Google's proprietary services. Unless it will integrate with some other kind of digital wallet.. probably open wallet foundation is a good start
> 
> Let's not get into useless discussions, almost nothing is impossible. Let's stay in a constructive discussion. Even SafetyNet and GCM are a sort of "proprietary services" but now they works.

My exact thoughts on this, I still think eventually we will get Wallet app fully working, and I'm hoping as soon as possible.

**Comment by apprehensions** [comment]
With MEETS_DEVICE_INTEGRITY, running google pay is stuck on the "Choose an account to use with Go...", and loops with this error in logcat:
```
04-19 11:02:01.589  9084  9100 W GoogleApiManager: yxv could not execute call because it requires feature (tapandpay_get_is_supervised_child_wallet_user, 1).
04-19 11:02:01.591  9084  9100 W GoogleApiManager: yxv could not execute call because it requires feature (tapandpay_get_is_supervised_child_wallet_user, 1).
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: Failed to get account feature
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: android.accounts.OperationCanceledException
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at android.accounts.AccountManager$Future2Task.internalGetResult(AccountManager.java:2606)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at android.accounts.AccountManager$Future2Task.getResult(AccountManager.java:2618)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at com.android.server.devicepolicy.DevicePolicyManagerService$CalculateHasIncompatibleAccountsTask.hasAccountFeatures(DevicePolicyManagerService.java:18815)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at com.android.server.devicepolicy.DevicePolicyManagerService$CalculateHasIncompatibleAccountsTask.userHasIncompatibleAccounts(DevicePolicyManagerService.java:18793)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at com.android.server.devicepolicy.DevicePolicyManagerService$CalculateHasIncompatibleAccountsTask.doInBackground(DevicePolicyManagerService.java:18781)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at com.android.server.devicepolicy.DevicePolicyManagerService$CalculateHasIncompatibleAccountsTask.doInBackground(DevicePolicyManagerService.java:18769)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at android.os.AsyncTask$3.call(AsyncTask.java:394)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at java.util.concurrent.FutureTask.run(FutureTask.java:317)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1156)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:651)
04-19 11:02:02.779  1399  3598 W DevicePolicyManager: 	at java.lang.Thread.run(Thread.java:1119)
```

Otherwise, clearing data, the google pay app is stuck on the home screen, similar to @KCGD.

**Comment by quimodotcom** [comment]
bump. this could be huge!!!!!!!

**Comment by wangtao99775** [comment]
Fix missing methods that cause application ANR #2925
Will this fix enable Google Wallet to run?

**Comment by ale5000-git** [comment]
> Fix missing methods that cause application ANR [#2925](https://github.com/microg/GmsCore/pull/2925) Will this fix enable Google Wallet to run?

It will likely fix the issue reported of @sewnie but there will likely be also other issues.

**Comment by sewnie** [comment]
Sorry, it isn't merged, despite being in a milestone. How can I test it?

**Comment by ale5000-git** [comment]
> Sorry, it isn't merged, despite being in a milestone. How can I test it?

It is now merged but you still have to wait microG v0.3.9 unless you compile it yourself.

**Comment by znbaboy** [comment]
anyone confirm if this works now? I figure you probably need to be rooted with Play Integrity Fix and Tricky Store to pass Integrity also, but are you able to add cards via GPay after that?

**Comment by sewnie** [comment]
Latest microG currently prevents anything other than BASIC from Play Integrity. Once that's fixed, perhaps this merged PR and STRONG integrity could make Google Pay work.

**Comment by znbaboy** [comment]
> Latest microG currently prevents anything other than BASIC from Play Integrity. Once that's fixed, perhaps this merged PR and STRONG integrity could make Google Pay work.

If that were the case, someone with just MicroG should be able to go through the card adding flow without any issue but get stopped when actually finalizing the adding of the card. They should then get a screen saying their card could not be added due to Play Integrity checks (rooted/modded device).

If someone can confirm that is what they are experiencing, then it is possible this will work now for people who are able to achieve device integrity through other means.

**Comment by quimodotcom** [comment]
Built microg, but experienced mass difficulty uninstalling and reinstalling new version, so much so i had to completely reflash mt device. So ima bump and hope someone lets me know when 0.3.9 is out.

**Comment by ale5000-git** [comment]
> Latest microG currently prevents anything other than BASIC from Play Integrity. Once that's fixed, perhaps this merged PR and STRONG integrity could make Google Pay work.

Maybe you didn't see this: https://github.com/microg/GmsCore/commit/2c55530c3d18266cd4f7e054f4f9722e6addd9b3

**Comment by sewnie** [comment]
That's great!! Yay!!! It would be nice if there were nightly releases signed with the key or a new release published.

**Comment by Iey4iej3** [comment]
It seems that there is no GitHub Action to produce test builds after every commit, say.

**Comment by quimodotcom** [comment]
Only issue with building it yourself is we dont have microg's official keypair. Meaning you have to completely uninstall microg then reinstall your new version. Apparently, theres a deadline for 0.3.9 for 24th july, so just have to wait to test then...

**Comment by quimodotcom** [comment]
Following the 0.3.9 prerelease, the only progress we have got is now you can no longer see your accounts in the selection page...

Edit: just to clarify, yes it infintely loads...

<img width="1240" height="2772" alt="Image" src="https://github.com/user-attachments/assets/5ee3bc78-3a26-4ef0-ac1c-b0014fd03094" />

**Comment by hongfa1380** [comment]
> > 最新的 microG 目前会阻止除 BASIC 之外的任何 Play Integrity 代码。修复此问题后，合并后的 PR 和 STRONG Integrity 或许能让 Google Pay 正常运行。
> 
> 如果真是这样，仅拥有 MicroG 的用户应该能够顺利完成卡添加流程，但在实际完成添加卡操作时会卡住。然后他们应该会看到一个屏幕，提示由于 Play Integrity 检查（已 root/修改的设备），无法添加他们的卡。
> 
> 如果有人可以确认这就是他们所经历的，那么对于能够通过其他方式实现设备完整性的人来说，这有可能起作用。

got strong pass also can't open wallet

**Comment by ale5000-git** [comment]
Can someone please post a logcat with microG v0.3.9?

**Comment by hongfa1380** [comment]
> Can someone please post a logcat with microG v0.3.9?

I will try. My device is pixel 6 with stock rom flash microG 0.3.9![PXL_20250801_071020961.RAW-01.MP.COVER.jpg](https://github.com/user-attachments/assets/bbe61f9a-af8e-4359-a269-ef0fcdefc7aa)

**Comment by hongfa1380** [comment]
> Can someone please post a logcat with microG v0.3.9?

[02_08-12-48-41_834.log](https://github.com/user-attachments/files/21556849/02_08-12-48-41_834.log)

**Comment by hongfa1380** [comment]
> Can someone please post a logcat with microG v0.3.9?

hello?

**Comment by rugk** [comment]
@hongfa1380 which "play integrity checker" app is that, which you are using?

**Comment by hongfa1380** [comment]
> [@hongfa1380](https://github.com/hongfa1380) which "play integrity checker" app is that, which you are using?

yes

**Comment by notpushkin** [comment]
@rugk [This one](https://play.google.com/store/apps/details?id=gr.nikolasspyr.integritycheck), judging by the icon

**Comment by rugk** [comment]
Thanks, it's apparently even (at leats the parts that can be) open-source and on GitHub: https://github.com/1nikolas/play-integrity-checker-app (/cc thanks @1nikolas for developing this!)

**Comment by 1nikolas** [comment]
Wait, Play Integrity has been bypassed? It was supposed to be unhackable.. Can this send to Google server whatever results we want? How did you even manage..

**Comment by KaKi87** [comment]
There are root-based options for this (I'm using one of those), although I don't know if they work with microG.

**Comment by hongfa1380** [comment]
> 等等，Play Integrity 被绕过了？它本来应该是无法破解的……这能把我们想要的结果发送到谷歌服务器吗？你们是怎么做到的……

mksu+ts+pif and key

**Comment by hongfa1380** [comment]
> 等等，Play Integrity 被绕过了？它本来应该是无法破解的……这能把我们想要的结果发送到谷歌服务器吗？你们是怎么做到的……

stock rom unlock bootloader mksu gki root

**Comment by hongfa1380** [comment]
> 等等，Play Integrity 被绕过了？它本来应该是无法破解的……这能把我们想要的结果发送到谷歌服务器吗？你们是怎么做到的……

need module and key

**Comment by hongfa1380** [comment]
![Image](https://github.com/user-attachments/assets/cd197bc3-4d83-4b21-9ce5-569b7a62c397)

nfc can't find google pay. So can't open google pay

**Comment by hongfa1380** [comment]
![Image](https://github.com/user-attachments/assets/d90ee2ec-bf48-4e41-8d8d-2b58d717ae3b)

**Comment by quimodotcom** [comment]
Im ngl, if I had the money, Id slap a bounty on this. This is one of the only remaining Google apps that havent even been slightly implemented. *by slightly implemented, I mean some sort of progress to a working app. the recent prerelease only included missing functions for the overall tapandpay*

**Comment by knocte** [comment]
I'd put a bounty if there was a reliable way to post it somewhere? it seems most bounty platforms have disappeared into the dust over the years.

**Comment by alexanmtz** [comment]
@knocte, this issue is on Gitpay: https://gitpay.me/#/profile/explore/task/797/android-pay-support

You can add a bounty there 👌

**Comment by knocte** [comment]
Are you sure GitPay is not dead? Look at their Team page: https://gitpay.me/#/team

**Comment by alexanmtz** [comment]
Yeah @knocte, is not dead, I'm the maintainer :)

**Comment by knocte** [comment]
>  Look at their Team page: https://gitpay.me/#/team

Nevermind, when I posted that, it wasn't loading properly, it looked like a page full or broken-link images.
However, GitPay claims to be part of https://worknenjoy.com/ and that indeed looks dead: blog page doesn't have any content, it claims only about $1,000 USD being exchanged (and about 1,000 tasks performed, interestingly the same number?).

I don't know what guarantee do I have that the dev would actually receive my bounty.

**Comment by alexanmtz** [comment]
@knocte, sorry for that. We can fix that, and our platform is fully open-source, so we have been working to address any issues.

Work n Enjoy is the company, and it could be a little outdated, it's just the company landing page, and the numbers you can see at the bottom of Gitpay:  _$2453 in bounties and freelancer work for 482 tasks to our community of 4063 users_

So we have more than 20k USD paid for developers. We will look over this and address the issues you pointed out.

When the developer sends the PR and merges it, he can claim. The payout will be sent automatically, and you can request a refund if it is not completed.

**Comment by quimodotcom** [comment]
> > Look at their Team page: https://gitpay.me/#/team
> 
> Nevermind, when I posted that, it wasn't loading properly, it looked like a page full or broken-link images. However, GitPay claims to be part of https://worknenjoy.com/ and that indeed looks dead: blog page doesn't have any content, it claims only about $1,000 USD being exchanged (and about 1,000 tasks performed, interestingly the same number?).
> 
> I don't know what guarantee do I have that the dev would actually receive my bounty.

Most of the GMSCore bountys are done through bountyhub if you wanna use them

**Comment by notpushkin** [comment]
@hongfa1380 do you have any updates? We’re all really excited :-)

**Comment by Aristag** [comment]
Any updates on this ?

**Comment by wangtao99775** [comment]
Is there any new news?

**Comment by Aristag** [comment]
Is There work being done for this???

**Comment by quimodotcom** [comment]
All ive found out is using a system root like SukiSu ultra, you can bypass Google Wallets Play Integrity lock howeverm you get stuck at a  "Not Implemented" screen tryna add a card.

**Comment by quimodotcom** [comment]
[https://i.postimg.cc/nX84xyNH/Screenshot-20260205-114044-Google-Wallet.png](https://i.postimg.cc/nX84xyNH/Screenshot-20260205-114044-Google-Wallet.png)
[https://i.postimg.cc/18pp3WZ6/Screenshot-20260205-114052-Google-Wallet.png](https://i.postimg.cc/18pp3WZ6/Screenshot-20260205-114052-Google-Wallet.png)

This is how far ive got. It used to say "... not implemented" on a blank screen but now it gets to the add to wallet screen and crashes.

**Comment by quimodotcom** [comment]
<img width="701" height="1568" alt="Image" src="https://github.com/user-attachments/assets/6568e4d8-3c78-4867-99f0-3a6701acf3fd" />

this is the main error im stuck on

**Comment by notpushkin** [comment]
> > > Look at their Team page: https://gitpay.me/#/team
> > 
> > Nevermind, when I posted that, it wasn't loading properly, it looked like a page full or broken-link images. However, GitPay claims to be part of https://worknenjoy.com/ and that indeed looks dead: [...]
> 
> Most of the GMSCore bountys are done through bountyhub if you wanna use them

That’s why I’ve just posted my bounty [here](https://github.com/microg/GmsCore/issues/361#issuecomment-1347431843) and not on any bounty platform :) I have _zero_ trust in any new players now, unfortunately.

microG could use some _reputable_ fiscal sponsor like the [Open Source Collective](https://opencollective.com/opensource) or [HCB](https://hackclub.com/fiscal-sponsorship/), and pay the bounties out from there? (Or reuse the [Liberapay account](https://liberapay.com/microG) for that, although IIRC it’s for recurring donations only, so it’ll be tricky to use it in that way.)

**Comment by Aristag** [comment]
Why so long for an update? Is this something that people don't won't or smth? Is it even being worked for the new microG update?

**Comment by notpushkin** [comment]
> Why so long for an update? Is this something that people don't won't or smth? Is it even being worked for the new microG update?

Keep in mind this is a volunteer-led project. That said, we’re all waiting for an update on this one, yeah :)

**Comment by Aristag** [comment]
So it IS being worked on? If so that's great news

**Comment by KaKi87** [comment]
> Keep in mind this is a volunteer-led project.

Still, at least one piece of opinion from one of the contributors would be nice : is it uninteresting to them, is it interesting but considered too hard, is it in the backlog, has it been discussed among them... Anything that would give any clue to whether this will ever exist.

**Comment by quimodotcom** [comment]
Basically, from my attempts, theres a few stubs and dummy implementations needed. however even once those are attempted (and my impls were shit), google wants strong integrity which to get microg on a rom, systemised and get strong integrity which usually requires locked bootloader, its incredibly challenging. im currently helping opstic get the RCS operational and fully tested but in the spare time, im currently waiting on an exploit for my phone model that will allow full locked bootloader spoof which will be a major help for this issue.

im as desperate for this as you lot, but itll take a WHILE since wallet wants the highest security unlike RCS 💀

**Comment by Aristag** [comment]
From my experience making wallet work isn't that hard. I managed to get it working with magisk modules and magiak alpha. My only problem is : Not yet supported: com.google.android.gms.pay.homescreen.additem METHOD

**Comment by quimodotcom** [comment]
> From my experience making wallet work isn't that hard. I managed to get it working with magisk modules and magiak alpha. My only problem is : Not yet supported: com.google.android.gms.pay.homescreen.additem METHOD

Same with me, its the systems after that aswell thatll require strong integrity which will be the harder challenge
